ghmirror/data_structures/redis_data_structures.py

```diff
@@ -0,0 +1,80 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2020
+# Author: Maha Ashour <mashour@redhat.com>
+
+"""
+Caching data in Redis.
+"""
+
+import os
+import pickle
+import redis
+
+
+class RedisCache:
+    """
+    Dictionary-like implementation for caching requests in Redis.
+    """
+    def __init__(self):
+        endpoint = os.environ.get('REDIS_ENDPOINT')
+        port = os.environ.get('REDIS_PORT', 6379)
+        self.cache = redis.Redis(
+            host=endpoint, port=port)
+
+    def __contains__(self, item):
+        sr_key = serialize(item)
+        return self.cache.exists(sr_key)
+
+    def __getitem__(self, item):
+        sr_key = serialize(item)
+        sr_value = self.cache.get(sr_key)
+        if sr_value is None:
+            raise KeyError(item)
+        return deserialize(sr_value)
+
+    def __setitem__(self, key, value):
+        sr_key = serialize(key)
+        sr_value = serialize(value)
+        self.cache.set(sr_key, sr_value)
+
+    def __iter__(self):
+        return self.scan_iter()
+
+    def scan_iter(self):
+        """
+        Make an iterator so that the client doesn't need to remember the cursor
+        position.
+        """
+        cursor = '0'
+        while cursor != 0:
+            cursor, data = self.cache.scan(cursor)
+            for item in data:
+                yield deserialize(item)
+
+    def __len__(self):
+        return self.cache.dbsize()
+
+    def __sizeof__(self):
+        return self.cache.info()['used_memory']
+
+
+def serialize(item):
```

Comment:
I'd place these two as private class methods. Or change the docstrings and move them to a more generic module, like `utils`.

Reply:
i had them as private methods initially but they got flagged in pr_check and i got a recommendation to change them to functions instead of methods since they never operate on the object itself. i think the correct approach would be to move them to a utils module then.

Reply:
You can just decorate them with @staticmethod

Reply:
just read up about the decorator and its exactly what i need thanks.